File: cpp/src/ch06/duct.cc
#include "monte_carlo.h"
#include "duct.h"

namespace duct
{

    Node::Node(const State &state) : state_(state),
                                     w_(0),
                                     n_(0) {}

    void Node::expand()
    {
        auto legal_actions0 = state_.legal_actions(0);
        auto legal_actions1 = state_.legal_actions(1);
        child_nodeses_.clear();
        for (const auto &action0 : legal_actions0)
        {
            child_nodeses_.emplace_back();
            auto &target_nodes = child_nodeses_.back();
            for (const auto &action1 : legal_actions1)
            {
                target_nodes.emplace_back(state_);
                auto &target_node = target_nodes.back();
                target_node.state_.advance(action0, action1);
            }
        }
    };

    double Node::explore()
    {

        if (state_.is_done())
        {
            double value = state_.white_score();
            w_ += value;
            n_++;
            return value;
        }

        if (child_nodeses_.empty())
        {
            State state_copy = state_;
            double value = playout(&state_copy);
            w_ += value;
            n_++;
            if (n_ == EXPAND_THRESHOLD)
                expand();
            return value;
        }

        // base has child nodeses
        double value = next_child_node().explore();
        w_ += value;
        n_++;
        return value;
    };

    double Node::t() const
    {
        // 分母の計算
        double t = 0;
        for (auto &child_nodes : child_nodeses_)
            for (auto &child_node : child_nodes)
                t += child_node.n_;
        return t;
    }

    double Node::ucb1(double w, double n) const
    {
        using std::log;
        using std::sqrt;
        return w / n + (double)C * sqrt(2. * log(t()) / n);
    }

    int Node::action0() const
    {
        double best_value = -INF;
        double best_i = -1;
        for (int i = 0; i < (int)child_nodeses_.size(); i++)
        {
            const auto &child_nodes = child_nodeses_[i];
            double w = 0;
            double n = 0;
            for (int j = 0; j < (int)child_nodes.size(); j++)
            {
                const auto &child_node = child_nodes[j];
                w += child_node.w_;
                n += child_node.n_;
            }
            double ucb1_value = ucb1(w, n);
            if (ucb1_value > best_value)
            {
                best_i = i;
                best_value = ucb1_value;
            }
        }
        return best_i;
    }

    int Node::action1() const
    {
        double best_value = -INF;
        double best_j = -1;
        for (int j = 0; j < (int)child_nodeses_[0].size(); j++)
        {
            double w = 0;
            double n = 0;
            for (int i = 0; i < (int)child_nodeses_.size(); i++)
            {
                const auto &child_node = child_nodeses_[i][j];
                w += child_node.w_;
                n += child_node.n_;
            }
            w = 1. - w;
            double ucb1_value = ucb1(w, n);
            if (ucb1_value > best_value)
            {
                best_j = j;
                best_value = ucb1_value;
            }
        }
        return best_j;
    }

    Node &Node::next_child_node()
    {
        // 0 のものは優先的に
        for (auto &child_nodes : child_nodeses_)
            for (auto &child_node : child_nodes)
                if (child_node.n_ == 0)
                    return child_node;

        // 0, 1 それぞれの最善を選ぶ
        int best_i = action0();
        int best_j = action1();

        return child_nodeses_[best_i][best_j];
    };

    int duct_action(const State &state,
                    const int player_id,
                    const int playout_number)
    {
        Node node = Node(state);
        node.expand();
        for (int i = 0; i < playout_number; i++)
            node.explore();

        auto legal_actions = state.legal_actions(player_id);
        int i_size = node.child_nodeses_.size();
        int j_size = node.child_nodeses_[0].size();
        // うまく 0,1 の切り替えできる方法ありそうだけど思いつかない

        if (player_id == 0)
        {
            int best_action_count = -1;
            int best_i = -1;
            for (int i = 0; i < i_size; i++)
            {
                int n = 0;
                for (int j = 0; j < j_size; j++)
                    n += node.child_nodeses_[i][j].n_;
                if (n > best_action_count)
                {
                    best_action_count = n;
                    best_i = i;
                }
            }
            return legal_actions[best_i];
        }
        else
        {
            int best_action_count = -1;
            int best_j = -1;
            for (int j = 0; j < j_size; j++)
            {
                int n = 0;
                for (int i = 0; i < i_size; i++)
                    n += node.child_nodeses_[i][j].n_;
                if (n > best_action_count)
                {
                    best_action_count = n;
                    best_j = j;
                }
            }
            return legal_actions[best_j];
        }
    }

}

File: rust/src/ch06/duct.rs
use crate::ch06::maze_state;

#[derive(Debug, Clone)]
struct DuctParams {
    c: f32,
    expand_threshold: usize,
}

#[derive(Debug, Clone)]
struct Node {
    n: usize,
    w: f32,
    child_nodeses: Vec<Vec<Node>>,
    state: maze_state::SimultaneousMazeState,
    params: DuctParams,
}

impl Node {
    fn new(
        state: &maze_state::SimultaneousMazeState,
        params: DuctParams,
    ) -> Self {
        let child_nodeses = Vec::new();
        Node {
            w: 0.0,
            n: 0,
            state: state.clone(),
            params,
            child_nodeses,
        }
    }

    fn increment(&mut self, value: f32) {
        self.w += value;
        self.n += 1;
    }

    fn t(&self) -> f32 {
        self.child_nodeses
            .iter()
            .flatten()
            .map(|node| node.n as f32)
            .sum()
    }

    fn ucb1(&self, w: f32, n: f32) -> f32 {
        w / n + self.params.c * (self.t().ln() / n).sqrt()
    }

    fn expand(&mut self) {
        self.child_nodeses.clear();
        let legal_actions0 = self.state.legal_actions(0);
        let legal_actions1 = self.state.legal_actions(1);
        for action0 in legal_actions0 {
            let mut child_nodes = Vec::new();
            for action1 in &legal_actions1 {
                let mut state = self.state.clone();
                state.advance(vec![action0, action1.to_owned()]);
                child_nodes.push(Node::new(&state, self.params.clone()));
            }
            self.child_nodeses.push(child_nodes);
        }
    }

    fn action0(&self) -> usize {
        // iterate i and sum j, than take ucb1
        let row_sums: Vec<f32> = self
            .child_nodeses
            .iter()
            .map(|row| {
                let w = row.iter().map(|node| node.w).sum::<f32>();
                let n = row.iter().map(|node| node.n).sum::<usize>() as f32;
                self.ucb1(w, n)
            })
            .collect();

        // get argmax
        row_sums
            .iter()
            .enumerate()
            .max_by(|(_, a), (_, b)| {
                a.partial_cmp(&b).unwrap_or(std::cmp::Ordering::Equal)
            })
            .map(|(index, _)| index)
            .unwrap()
    }

    fn action1(&self) -> usize {
        // transpose
        let transposed: Vec<Vec<&Node>> = (0..self.child_nodeses[0].len())
            .map(|i| self.child_nodeses.iter().map(|row| &row[i]).collect())
            .collect();

        // take col sum
        let col_sums: Vec<f32> = transposed
            .iter()
            .map(|col| {
                let w = 1.0 - col.iter().map(|node| node.w).sum::<f32>();
                let n = col.iter().map(|node| node.n).sum::<usize>() as f32;
                self.ucb1(w, n)
            })
            .collect();

        // get argmax
        col_sums
            .iter()
            .enumerate()
            .max_by(|&(_, a), &(_, b)| {
                a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal)
            })
            .map(|(index, _)| index)
            .unwrap()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn setup() -> Node {
        let state_params = maze_state::MazeParams {
            height: 3,
            width: 3,
            end_turn: 4,
        };
        let state = maze_state::SimultaneousMazeState::new(0, state_params);
        let params = DuctParams {
            c: 1.0,
            expand_threshold: 3,
        };
        Node::new(&state, params)
    }

    #[test]
    fn test_action_1() {
        let mut node = setup();
        node.expand();
        node.child_nodeses[0][0].n = 1;
        node.child_nodeses[0][0].w = 0.5;
        node.child_nodeses[1][1].n = 1;
        node.child_nodeses[1][1].w = 0.5;
        node.child_nodeses[2][2].n = 1;
        node.child_nodeses[2][2].w = 0.7;
        assert_eq!(node.action1(), 1);
    }

    #[test]
    fn test_action_0() {
        let mut node = setup();
        node.expand();
        node.child_nodeses[0][0].n = 1;
        node.child_nodeses[0][0].w = 0.5;
        node.child_nodeses[1][1].n = 1;
        node.child_nodeses[2][2].n = 1;
        assert_eq!(node.action0(), 0);
    }

    #[test]
    fn test_expand() {
        let mut node = setup();
        assert!(node.child_nodeses.is_empty());
        node.expand();
        assert!(!node.child_nodeses.is_empty());
        assert!(!node.child_nodeses[0].is_empty());
    }

    #[test]
    fn test_ucb1() {
        let mut node = setup();
        node.n = 1;
        let child_nodes = vec![node.clone(); 3];
        let child_nodeses = vec![child_nodes; 3];
        node.child_nodeses = child_nodeses;
        let actual = node.ucb1(2.0, 3.0);
        assert_eq!(actual, 1.5224752);
    }

    #[test]
    fn test_t() {
        let mut node = setup();
        node.n = 1;
        let child_nodes = vec![node.clone(); 3];
        let child_nodeses = vec![child_nodes; 3];
        node.child_nodeses = child_nodeses;
        let actual = node.t();
        assert_eq!(actual, 9.0);
    }

    #[test]
    fn test_increment() {
        let mut node = setup();
        node.increment(0.5);
        assert_eq!(node.n, 1);
        assert_eq!(node.w, 0.5);
    }

    #[test]
    fn make_node() {
        let node = setup();
        assert!(node.child_nodeses.is_empty());
    }
}
